### PR TITLE
Add partitioned database support.

### LINKED ISSUE #372: Add partitioned database support.
Implement partitioned database support in nodejs-cloudant.

API reference: https://cloud.ibm.com/docs/services/Cloudant/guides?topic=cloudant-database-partitioning#partitioned-databases

See similar feature PRs for [python-cloudant][0] and [java-cloudant][1].

[0]: https://github.com/cloudant/python-cloudant/pull/441
[1]: https://github.com/cloudant/java-cloudant/pull/473

### PR DESCRIPTION
**Pull request by smithsz**
## Checklist

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)
- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`|`CHANGELOG.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description
Add partitioned database support.

Fixes #372.

## Approach
Add additional partitioned functions to the `cloudant` object.

## Schema & API Changes
Adds new API:

```js
const db = cloudant.db.use(dbName);

// partition info
db.partitionInfo(partitionKey).then((body) => { ... });

// partitoned find
db.partitionedFind(partitionKey, selector, callback).then((body) => { ... });
db.partitionedFindAsStream(partitionKey, selector).pipe( ... )

// partitioned all docs
db.partitionedList(partitionKey, callback).then((body) => { ... });
db.partitionedListAsStream(partitionKey).pipe( ... )

// partitioned search
db.partitionedSearch(partitionKey, designName, searchName, params, callback).then((body) => { ... });
db.partitionedSearchAsStream(partitionKey, designName, searchName, params).pipe( ... )

// partitioned view
db.partitionedView(partitionKey, designName, viewName, params, callback).then((body) => { ... });
db.partitionedViewAsStream(partitionKey, designName, viewName, params).pipe( ... )
```

## Security and Privacy
No change.

## Testing
Includes additional unit tests.

## Monitoring and Logging
No change.

### CODE REVIEW

File: cloudant.js [hunk 1]
```diff
@@ -75,6 +86,15 @@ function Cloudant(options, callback) {
   // Cloudant Database Functions
   // ===========================
 
+  // https://cloud.ibm.com/docs/services/Cloudant/guides?topic=cloudant-database-partitioning#creating-the-database
+  function createPartitioned(db, callback) {
```
Comment (emlaver) [inline cloudant.js]: Can we not just support the `partitioned` parameter in `db.create` like you did for `python-cloudant`?

File: cloudant.js [hunk 2]
```diff
@@ -25,6 +25,17 @@ const Client = require('./lib/client.js');
 // return a URL.
 var reconfigure = require('./lib/reconfigure.js');
 
+function getCallback(opts, callback) {
```
Comment (emlaver) [inline cloudant.js]: Can you add a one or two line comment about why this function was added?

File: cloudant.js (hunk 1 above)
Comment (smithsz) [reply cloudant.js]: yeah, can do. how about `nano.db.create(name, [params], [callback])`? i.e.
```js
nano.db.create('mydb', { partitioned: true }, cb);
```

Comment (emlaver) [reply cloudant.js]: That looks good to me

File: test/partitioned_databases.js [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        .map(() => { return uuidv4(); });
+
+  before(function(done) {
+    var mocks = nock(SERVER)
+      .put(`/${DBNAME}`)
+      .query({ partitioned: true })
+      .reply(201, { ok: true });
+
+    const cloudant = Cloudant({ url: SERVER, username: ME, password: PASSWORD });
+    cloudant.db.create(DBNAME, { partitioned: true }).then((body) => {
```
Comment (ricellis) [inline test/partitioned_databases.js]: The test style is fine, but if we wanted to make it more concise now that we have promises I think we can just return the promise to mocha i.e.
```js
return cloudant.db.create(DBNAME, { partitioned: true }).then((body) => {
      assert.ok(body.ok);
      mocks.done();
    })
```

File: test/partitioned_databases.js [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    cloudant.db.destroy(DBNAME).then((body) => {
+      assert.ok(body.ok);
+      mocks.done();
+      done();
+    }).catch(done);
+  });
+
+  it('create a partitioned database', function(done) {
+    var mocks = nock(SERVER)
+      .get(`/${DBNAME}`)
```
Comment (ricellis) [inline test/partitioned_databases.js]: Given the test is for creating, shouldn't this be a `PUT`? Or I guess actually we are asserting that the `before` create worked, maybe `created a partitioned database` is a better test name then.

File: test/partitioned_databases.js [hunk 5]
```diff
@@ -78,7 +75,7 @@ describe('Partitioned Databases #db', function() {
     const db = cloudant.db.use(DBNAME);
 
     var q = async.queue(function(task, callback) {
-      db.bulk({ 'docs': task.docs }).then(callback).catch(done);
```
Comment (ricellis) [inline test/partitioned_databases.js]: Don't we still need this one, since we aren't returning the promise here?

**Pull request merged**